File: cjc/POWSUM-solutions/PowSumNT.java
import java.io.*;

public class PowSumNT {
   final static int P = 1000000007;

   static int powMod(long x, int n, int M) {
      long res = 1;
      x %= M;
      while (true) {
         if ((n & 1) != 0)
            res = (res * x) % M;
         n /= 2;
         if (n == 0) break;
         x = (x * x) % M;
      }
      return (int) res;
   }

   int S(long x, int n) {
      // x==1  => no modular inverse, but S(1, n) is trivial
      if (x == 1) return n+1;

      long a = (1L -x + P) % P;
      // a must be co-prime of M
      int inv = powMod(a, P-2, P);
      int xn1 = powMod(x, n+1, P);
      long res = ((1L - xn1 + P) * inv) % P;
      return (int) res;
   }

   public static void main (String[] args) throws IOException {
      (new PowSum()).Run();
   }

   void Run() throws IOException {
      BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
      int TC = Integer.parseInt(in.readLine());
      for (int tc = 1; tc <= TC; ++tc) {
         String[] tok = in.readLine().split(" ");
         int x = Integer.parseInt(tok[0]);
         int n = Integer.parseInt(tok[1]);
         int res = S(x, n);
         System.out.println(res);
      }
   }
}

File: cjc/POWSUM-solutions/powsum_nt.cpp
#include <iostream>

using namespace std;

/*
http://en.wikipedia.org/wiki/Geometric_progression#Geometric_series
  S(x, n) = (1 - x^(n+1)) / (1-x)

Fermat's Little Theorem
  a^(p-1) = 1 (mod p)
  a^(p-2) * a = 1 (mod p)
  a^(p-2) is modular inverse of a (mod p)  if p is prime
*/

const unsigned int P = 1000000007;

typedef unsigned long long ullong;
typedef unsigned int uint;

uint powmod(ullong x, uint n, uint M) {
   ullong res = 1;
   x %= M;
   while (true) {
      if (n & 1) {
         res = (res * x) % M;
      // n--;
      }
      n >>= 1;
      if (n == 0) break;
      x = (x * x) % M;
   }
   return (uint) res;
}

uint S(ullong x, uint n) {
   // case where x == 1 leads to non existence of inverse
   //   fortunately, S(1, n) == n+1 (easy to compute)
   if (x == 1) return n+1;

   ullong a = (1LL - x + P) % P;
   // a must be co-prime of M
   uint inv = powmod(a, P-2, P);
   uint xn1 = powmod(x, n+1, P);
// cout << "a = " << a << "  inv = " << inv << "  rn1 = " << rn1 << endl;
// avoid overflow by converting to 64-bit integer before multiplication
   return (ullong(1LL - xn1 + P) * inv) % P;
}

int main(int argc, char* argv[]) {
   ios_base::sync_with_stdio(false);
   int TC;
   cin >> TC;
   for (int tc = 1; tc <= TC; ++tc) {
      int x, n;
      cin >> x >> n;
      uint res = S(x, n);
      cout << res << endl;
   }
   return 0;
}
